pkg/util/taskgroup/taskgroup.go

```diff
@@ -0,0 +1,132 @@
+// Originally taken from https://github.com/ptxmac/multierrgroup
+
+// Package taskgroup provides a mix of multierr and errgroup
+// See documentation for https://pkg.go.dev/go.uber.org/multierr and https://pkg.go.dev/golang.org/x/sync/errgroup
+package taskgroup
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+
+	"github.com/neondatabase/autoscaling/pkg/util/stack"
+)
+
+// Group manages goroutines and collect all the errors.
+// See https://pkg.go.dev/golang.org/x/sync/errgroup#group for more information
+type Group interface {
+	Ctx() context.Context
+	WithPanicHandler(f func(any))
+	Wait() error
+	Go(name string, f func(logger *zap.Logger) error)
+}
+
+type group struct {
+	cancel       context.CancelFunc
+	ctx          context.Context
+	logger       *zap.Logger
+	panicHandler func(any)
+
+	wg sync.WaitGroup
+
+	errMutex sync.Mutex
+	err      error
+}
+
+type GroupOption func(*group)
+
+// WithParentContext sets the parent context for the group.
+func WithParentContext(ctx context.Context) GroupOption {
+	return func(g *group) {
+		g.ctx, g.cancel = context.WithCancel(ctx)
+	}
+}
+
+// NewGroup returns a new Group.
+func NewGroup(logger *zap.Logger, opts ...GroupOption) Group {
+	g := &group{
+		cancel:       nil, // Set separately by Ctx
+		ctx:          nil, // Set separately by Ctx
+		panicHandler: nil, // Set separately by WithPanicHandler
+		logger:       logger,
+		wg:           sync.WaitGroup{},
+
+		errMutex: sync.Mutex{},
+		err:      nil,
+	}
+
+	for _, opt := range opts {
+		opt(g)
+	}
+	if g.ctx == nil {
+		// If parent context is not set, use background context
+		WithParentContext(context.Background())(g)
+	}
+
+	return g
+}
+
+// Ctx returns a context that will be canceled when the group is Waited.
+func (g *group) Ctx() context.Context {
+	return g.ctx
+}
+
+// WithPanicHandler sets a panic handler for the group.
+func (g *group) WithPanicHandler(f func(any)) {
```

Comment:
Seems like this should also be a `GroupOption`?

Reply:
Changed in 75dc6de